### ISSUE TITLE
List registered events: The `events` builtin

### ISSUE THREAD

**Issue by derekstavis** [fish-shell/fish-shell#2640]
Currently there is a way to register and emit events, but there isn't a way to list events that functions registered to, neither there are ways to unregister from events, aside from removing functions. I don't currently see the need to unregister feature, but listing registered events would be useful for debugging.

Maybe it's time to add a `events` builtin, analogous to `functions` one:

```
$ events --help

       events ‐‐ print, query or remove events

   Synopsis
       events [ ‐a | ‐‐all ] [ ‐n | ‐‐names ]
       events [ ‐e | ‐q ] EVENTS...

   Description
       events prints, queries or remove events.

       The following options are available:

       · ‐a or ‐‐all lists all events, even those whose name start with an underscore.

       · ‐n or ‐‐names lists the names of all defined events.

       · ‐e or ‐‐erase causes the specified event handlers (functions) to be erased.

       · ‐q or ‐‐query tests if the specified events exist.

```

Ideas, opinions on the matter?

**Comment by krader1961** [comment]
Is that level of flexibility needed, especially for something that is primarily a diagnostic tool? I'd prefer a new option for the __functions__ builtin that simply shows the events and associated functions rather than defining a new builtin. Perhaps __-E__ and __--list-event-handlers__ options. It would also be helpful to add __-V__ and __--list-variable-handlers__ options.

Also, your __--query__ option confuses me.  What exactly does that do? There are predefined events but you can define and use arbitrary event names (see `man emit`). So what does it me to test if the specified event exists? Same for the __--name__ option.

**Comment by derekstavis** [comment]
>Is that level of flexibility needed, especially for something that is primarily a diagnostic tool?

I'm not sure. I just wanted to promote events in the same level as functions with this idea (it's a mostly copy-paste of `functions` man page).

>Also, your **--query** option confuses me. What exactly does that do?
>(...) but you can define and use arbitrary event names

`-q` looks like a real use case to me: Say one wants to check if a event handler was already defined before attaching a handler:

```
if not events -q some_event
  function __some_event_handler --on-event some_event
    echo foobar
  end
end
```

>Same for the **--name** option.

Yeah, I agree with in this case, but, as I said, the proposition was mostly a copy-paste from `functions`.

>(...) I'd prefer a new option for the functions builtin

We currently don't have a bidirectional relationship between functions and events, in a way that we can't know what functions are mapped to what events, so erasing or deassociating events are not currently possible. I think that a new builtin for events would make this mapping more clear.

**Comment by krader1961** [comment]
I'm relatively new to fish but what exactly do you mean by "erasing events"? There are predefined events (e.g., __fish_prompt__). What would it mean to erase such an event? Ignoring the predefined events event names are arbitrary strings that do not need to be defined. You can run `emit arglebargle` and any function that registered an interest in that event name will be run (and if no function is interested in the event the `emit` simply does nothing).

I'm also unconvinced the __-q__ option is useful. It is perfectly reasonable, and legal, to define multiple functions that are run when an event is emitted.

I agree that it would be nice if there were a way to display which functions registered an interest in events and changes to variables but wanting "to promote events in the same level as functions" doesn't seem to make any sense.

**Comment by derekstavis** [comment]
>(...) what exactly do you mean by "erasing events"?

Erasing all event handlers from a given event name.

>I'm also unconvinced the -q option is useful. It is perfectly reasonable, and legal, to define multiple functions that are run when an event is emitted.

What if you want to define just **one** handler for a given event?

>but wanting "to promote events in the same level as functions" doesn't seem to make any sense.

What I mean here is that events are somewhat hidden into `function` and `emit` builtins. Promoting them "in the same level as `functions`" means adding a builtin to deal with events. I also don't think that this kind of operation is `functions` builtin duty.

**Comment by faho** [comment]
>Erasing all event handlers from a given event name.

This currently has a problem with autoloading - if the event handler, specified in a function file (in $fish_function_path) hasn't been executed once, fish won't know that it even exists. Not that this kills the idea - see the next paragraph for an example. It just means that it's a lot more specific in use than users would likely expect, and changing that would require at least weakening the auto-loading mechanic (by peeking inside the files) if not wholly abandoning it.

>What if you want to define just one handler for a given event?

command-not-found-handling is of interest here, though we'd do one or the other. If the current optimization (which I'm not convinced actually optimizes anything) of loading the real command-not-found handler only the first time the event is fired stays, then we'd want to do this - `events -q fish_command_not_found; and return 0`. In that case the user would need to define a handler (even if they don't want any handling, they'd have to have a stub) to inhibit the ones shipped with fish (the difference to the current situation being that not the function name matters, but that it listens to this event). If, as I suspect, the optimization is useless (I can't measure any difference), then we'd tell users to use `events --erase fish_command_not_found` (which could also look like `functions --erase (events --list fish_command_not_found)`).

I do believe that it should be clearer what in this proposal applies to _events_, and what to event _handlers_ - like the "--erase" option. It can't erase an event because it is created by emitting. It can erase the currently registered event handlers, though.

**Comment by derekstavis** [comment]
>command-not-found-handling is of interest here (...)

This is one of the use cases inside fish shell. And it's very useful for frameworks.

>(...) I do believe that it should be clearer what in this proposal applies to events, and what to event handlers

I think that when I say events I really mean handlers.

The whole idea of the proposal is to do the reverse mapping between event names and handlers, allowing operate over **event handlers** via **event names**. Minimum operations include listing, querying (as seen in `set -q` and `functions -q`) and erasing **handlers** by event **name**.

**Comment by krader1961** [comment]
We should implement a way to list functions that monitor events or variable changes along with the names of the events or variables being monitored. For debugging if nothing else. We should not implement a feature to stop all functions from monitoring an event, let alone erasing those functions, by event name. If the former capability, listing functions listening for a given change, is implemented then it becomes trivial to selectively remove such a function. That then makes it possible for someone who really wants to remove all functions registering an interest in the event to do so. But doing so is dangerous and should not be supported by default.

**Comment by jchilders** [comment]
#4694 looks to cover similar functionality, although I prefer the solution proposed here.

**Comment by faho** [comment]
We have `functions --handlers` for showing event handlers, and registering/unregistering isn't a *thing* because events are simply created when they are fired.

Fixed in fish 3.0 with the introduction of `functions --handlers`.
